docs/architecture/adr-047-extend-upgrade-plan.md

```diff
@@ -0,0 +1,185 @@
+    1.  If a `checksum` is provided (in the field or as a query param), the checksum of the result of the `url` MUST equal the provided checksum.
+
+### Cosmovisor updates
+
+If the `upgrade-info.json` file does not contain any `UpgradeInstructions`, existing functionality will be maintained.
+
+We will update Cosmovisor to look for and handle the new `UpgradeInstructions` in `upgrade-info.json`.
+If the `UpgradeInstructions` are provided, we will do the following:
+1.  The `info` field will be ignored.
+1.  The `artifacts` field will be used to identify the artifact to download based on the `platform` that Cosmovisor is running in.
+1.  If a `checksum` is provided (either in the field or as a query param in the `url`), and the downloaded artifact has a different checksum, the upgrade process will be interrupted and Cosmovisor will exit with an error.
+1.  If a `pre_run` command is defined, it will be executed at the same point in the process where the `app pre-upgrade` command would have been executed.
+    It will be executed using the same environment as other commands run by Cosmovisor.
+1.  If a `post_run` command is defined, it will be executed after executing the command that restarts the chain.
+    It will be executed in a background process using the same environment as the other commands.
+    Any output generated by the command will be logged.
+    Once complete, the exit code will be logged.
+
+## Consequences
+
+### Backwards Compatibility
+
+Since the only change to existing definitions is the addition of the `upgrade` field to the `Plan` message, and that field is optional, there are no backwards incompatibilities with respects to the proto messages.
+Additionally, current behavior will be maintained when no `UpgradeInstructions` are provided, so there are no backwards incompatibilities with respects to either the upgrade module or Cosmovisor.
+
+### Forwards Compatibility
+
+In order to utilize the `UpgradeInstructions` as part of a software upgrade, both of the following must be true:
+1.  The chain must already be using a sufficiently advanced version of the Cosmos SDK.
+1.  The chain's nodes must be using a sufficiently advanced version of Cosmovisor.
+
+### Positive
+
+1.  The structure for defining artifacts is clearer since it is now defined in the proto instead of in documentation.
+1.  Availability of a pre-run command becomes more obvious.
+1.  A post-run command becomes possible.
+
+### Negative
+
+1.  The `Plan` message becomes larger.
+1.  There is no option for providing a URL that will return the `UpgradeInstructions`.
+1.  The only way to provide multiple assets for a platform is to use an archive as the platform's artifact.
+1.  A chain software upgrade is required before this new functionality can be used.
+    I.e. The next upgrade must use the old definitions, but the one after that can use the new structure.
```

Comment:
I'm not sure if this is negative consequence: every feature which touches the state changes the state machines, so it's _breaking_.

Reply:
I included this as a negative because it's possible to design the Cosmovisor changes so that this isn't the case.

Specifically, update Cosmovisor to further overload the `info` field to allow either the JSON of this new `UpgradeInstructions` or a url that returns that JSON (as well as what's currently looked for there). Then, as long as the node is running the new version of Cosmovisor, the new functionality from ADR is available on the next chain upgrade.

I will add that to the "Cosmovisor Updates" section and then remove this as a negative.

Reply:
Removed this negative and added to the "Cosmovisor Updates" section, lines 134-136.